### PR TITLE
#164069215 JWT Auth on successful login and registration.

### PR DESCRIPTION
**Pull request by BagzieGracious**
#### What does this PR do?
* This PR enables users receive a JWT upon successful Registration and Login, and then implements the newly strategy for authentication.

#### Description of Task to be completed?
* Decoded user object into a token.
* Return token to successfully logged in user.
* Implemented JWTAuthentications in backends.
* Added token field in models.

#### How should this be manually tested?
* Clone the repo and cd into Ah-backend-aquaman
* git checkout ft-jwt-authentication-164069215
* pip install -r requirements.txt
* python manage.py makemigrations
* python manage.py migrate
* python manage.py runserver
 
The below endpoints will be used to log in with
* Registration endpoint
`POST http://127.0.0.1:8000/api/users/`
* Login endpoint
`POST http://127.0.0.1:8000/api/users/login/`

#### Any background context you want to provide?
* N/A

#### What are the relevant pivotal tracker stories?
[#164069215](https://www.pivotaltracker.com/story/show/164069215)

#### screenshots
* Registration screenshot
screenshot 2019-03-04 at 17 25 45

* Login screenshot
screenshot 2019-02-28 at 20 40 35

### CODE REVIEW

**Review by joelethan — COMMENTED** [review]
@BagzieGracious I think you should add an image to show how the user inputs their credentials when logging in and registering

File: authors/apps/authentication/backends.py [hunk 1]
```diff
@@ -1,9 +1,35 @@
...
+            return self.authentication_credentials(request, token)
+        return None
+
+    def authentication_credentials(self, request, token):
+        """returns the decoded token"""
+        try:
+            payload = jwt.decode(token, settings.SECRET_KEY)
+            user = User.objects.get(email=payload['email'])
+            if not user.is_active:
+                raise exceptions.AuthenticationFailed("User was deactivated.")
```
Comment (CryceTruly) [inline authors/apps/authentication/backends.py]: @BagzieGracious thanks for the good work, I think we can make this error message more descriptive to the user so that they know what to do after, something like 
```
Your account was deactivated
```
Then with followup instructions on how to re-activate it.

File: authors/apps/authentication/backends.py [hunk 2]
```diff
@@ -1,9 +1,35 @@
...
+            return (user, token)
+        except:
+            raise exceptions.AuthenticationFailed("Your token is invalid or is expired.")
```
Comment (CryceTruly) [inline authors/apps/authentication/backends.py]: @BagzieGracious , thanks for the good work, however, I think we could we could tell a user what to do next.

**Comment by Peace-Apple** [comment]
Good work, could you also add a screenshot showing some of these things in action, like when someone logs in or signs up

**Review by Peace-Apple — DISMISSED** [review]
LGTM

**Comment by CryceTruly** [comment]
LGTM

**Review by joelethan — DISMISSED** [review]
LGTM

**Review by wasibani-roy — DISMISSED** [review]
LGTM cloned and run it successfully on my computer

**Review by CryceTruly — DISMISSED** [review]
LGTM

**Review by Peace-Apple — DISMISSED** [review]
LGTM

**Review by joelethan — DISMISSED** [review]

**Review by marthamareal — COMMENTED** [review]
Good work @BagzieGracious , You can also add tests for the different edge cases for example `what happens when an invalid token is provided`

**Review by CryceTruly — CHANGES_REQUESTED** [review]
Thanks @BagzieGracious for the good work, you just have to make sure you test enough as your push decreased test coverage by 4%.

**Review by joelethan — DISMISSED** [review]
@BagzieGracious take a look at the ReadMe.md it has some repetitive information

**Review by CryceTruly — DISMISSED** [review]
LGTM

**Review by wasibani-roy — DISMISSED** [review]
Good work @BagzieGracious you where able to add tests however you could consider adding a minimum length to your password field under your login serializer just to ensure a reasonable password is entered

**Review by wasibani-roy — APPROVED** [review]
LGTM good job @BagzieGracious

**Review by CryceTruly — APPROVED** [review]
LGTM

**Review by joelethan — APPROVED** [review]
